Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean

```diff
@@ -172,17 +172,55 @@ instance instDecidableMemSupp (c : G.ConnectedComponent) (v : V) : Decidable (v
   c.recOn (fun w ↦ decidable_of_iff (G.Reachable v w) $ by simp)
     (fun _ _ _ _ ↦ Subsingleton.elim _ _)
 
+lemma ConnectedComponent.odd_card_supp_iff_odd_subcomponents {G'} [DecidableRel G.Adj]
+    (h : G ≤ G') (c' : ConnectedComponent G') :
+    Odd (Nat.card c'.supp) ↔ Odd (Nat.card
+    ({c : ConnectedComponent G | c.supp ⊆ c'.supp ∧ Odd (Nat.card c.supp) })) := by
+  haveI : DecidablePred (fun c : ConnectedComponent G ↦ c.supp ⊆ c'.supp) := Classical.decPred _
+  haveI : (i : G.ConnectedComponent) → Fintype ↑(⋃ (_ : i.supp ⊆ c'.supp), i.supp) := by
+    exact fun i ↦ Set.fintypeiUnion fun _ ↦ i.supp
+  nth_rewrite 1 [← (c'.biUnion_supp_eq_supp h)]
+  rw [@Nat.card_eq_card_toFinset, @Set.toFinset_iUnion _ (ConnectedComponent G) _ _
+    (fun c => ⋃ (_ : c.supp ⊆ c'.supp), c.supp),
+    Finset.card_biUnion (by
+    intro x _ y _ hxy
+    simp only [Set.disjoint_toFinset, Set.disjoint_iUnion_right, Set.disjoint_iUnion_left]
+    exact fun _ _ ↦ pairwise_disjoint_supp_connectedComponent _ hxy
+    )]
+  simp only [Set.toFinset_card]
+  rw [Finset.odd_sum_iff_odd_card_odd, Nat.card_eq_fintype_card, Fintype.card_ofFinset,
+    Finset.filter_congr (by
+    intro x _
+    constructor <;> intro h
+    · have := Odd.pos h
+      rw [@Fintype.card_pos_iff, Set.nonempty_coe_sort, Set.nonempty_iUnion] at this
+      obtain ⟨hs, _⟩ := this
+      refine ⟨hs, ?_⟩
+      rw [← @Set.toFinset_card] at h
+      haveI : Nonempty (x.supp ⊆ c'.supp) := Nonempty.intro hs
+      simp_rw [Set.iUnion_const] at h
+      rw [@Nat.card_eq_card_toFinset]
+      exact h
+    · haveI : Nonempty (x.supp ⊆ c'.supp) := Nonempty.intro h.1
+      simp_rw [Set.iUnion_const, Fintype.card_eq_nat_card]
+      exact h.2 : ∀ x ∈ (Finset.univ : Finset G.ConnectedComponent),
+      Odd (Fintype.card ↑(⋃ (_ : x.supp ⊆ c'.supp), x.supp)) ↔
+      (x ∈ {c : ConnectedComponent G | c.supp ⊆ c'.supp ∧ Odd (Nat.card c.supp)}))]
+
 lemma odd_card_iff_odd_components : Odd (Nat.card V) ↔
     Odd (Nat.card ({(c : ConnectedComponent G) | Odd (Nat.card c.supp)})) := by
-  rw [Nat.card_eq_fintype_card]
-  simp only [← (set_fintype_card_eq_univ_iff _).mpr G.iUnion_connectedComponentSupp,
-    ConnectedComponent.mem_supp_iff, Fintype.card_subtype_compl,
-    ← Set.toFinset_card, Set.toFinset_iUnion ConnectedComponent.supp]
-  rw [Finset.card_biUnion
-    (fun x _ y _ hxy ↦ Set.disjoint_toFinset.mpr (pairwise_disjoint_supp_connectedComponent _ hxy))]
-  simp_rw [Set.toFinset_card, ← Nat.card_eq_fintype_card]
-  rw [Nat.card_eq_fintype_card, Fintype.card_ofFinset]
-  exact (Finset.odd_sum_iff_odd_card_odd (fun x : G.ConnectedComponent ↦ Nat.card x.supp))
+  cases isEmpty_or_nonempty V with
```

Comment:
I guess here I meant `cases'` since we don't want to name the hypothesis
```suggestion
  cases' isEmpty_or_nonempty V
```

Reply:
Changed, do I didn't know how to do the second part without naming the hypothesis, so I added a `with` to it